pokt-network/pocket-core
# Allow the output address owner to change the output address via MsgStake

This patch allows the owner of the output address of a staked node to change the output address to another via the existing stake transaction.  The same idea was proposed as #1499.

### Rationale:

This feature gives users more flexibility.  One of the possible use cases is that a node runner can convert custodial stake nodes into non-custodial stake nodes without unstaking.  Or a current non-custodial stake user may want to change the output address some time after they stake nodes.

### Technical details:

Below are the key functions in the transaction processing flow.  This patch makes a new `MsgStake` transaction to change the output address pass through all Validate* functions and lets the final function `EditStakeValidator` accept the change in the world state if the new feature key “OEDIT” is enabled.

1. BaseApp.runTx (baseapp.go)
    1. auth.ValidateTransaction (ante.go)
    2. BaseApp.runMsg (baseapp.go)
        1. nodes.handleStake (handler.go)
            1. keeper.ValidateValidatorStaking (valStateChanges.go)
                1. keeper.ValidateValidatorMsgSigner (for the new validator)
                2. keeper.ValidateValidatorMsgSigner (for the current validator)
                3. keeper.ValidateEditStake
            2. keeper.StakeValidator
                1. keeper.EditStakeValidator

The challenge was in `ValidateTransaction`, which expects a transaction message to be signed by a key which appears in the transaction itself.  For example, `MsgStake` must be signed by `MsgStake.PublicKey` or `MsgStake.Output`.  Since we intend to allow only the current output address owner to change the output address, however, a new `MsgStake` transaction must be signed by the current output address, and it doesn’t appear in the transaction because a new output address is there.  This patch needs to change that signer restriction rule.

Each message type implements `GetSigners` that returns a slice of `Address`.  When a transaction arrives at `ValidateTransaction`, it calls `GetSigners` and checks if a transaction is signed by any of the addresses returned from `GetSigners`.  If not, the transaction is rejected with `CodeUnauthorized`.  The challenge here is that `ValidateTransaction` is in the auth package that doesn’t know about `MsgStake`.  One possible solution is to make `MsgStake.GetSigners` return three addresses, `MsgStake.PublicKey`, `MsgStake.Output`, and the current output address.  However, implementing it is not straightforward because `MsgStake` is in the types package which doesn’t have direct access to the keeper package that is required to get the current output address.

The proposed solution in this patch is to introduce a new method `GetMsgStakeOutputSigner` in the keeper package and `ValidateTransaction` calls it to append the current output address to the slice returned from `GetSigners`.

The changes in valStateChanges.go are straightforward.  `ValidateValidatorMsgSigner` simply checks if a signer is either the node’s operator or output address.  We need to skip the call to `ValidateValidatorStaking` with the new validator because the signer does not appear in the new validator.  The final validate function `ValidateEditStake` ensures that `MsgStake` to change the output address must be signed by the current output address.  Otherwise it returns a new error `ErrDisallowedOutputAddressChange`.

After all validate functions, `EditStakeValidator` accepts a new output address.

### Testing:

1. Prepare a local network with NCUST enabled.
2. Set the following environment variables accordingly.
    ```
    export POKT=<path to pocket executable>
    export CHAINS=<chain IDs to host>
    export URL=<service url>
    export STAKE=<upokt to stake>
    export NETWORK=<network name>
    export NODE=<node address>
    export PUBKEY=<node public key>
    export PRIVKEY=<node private key>
    export OUTPUT1=<output address>
    export OUTPUT1_PRIVKEY=<OUTPUT1’s private key>
    export OUTPUT2=<another output address>
    export OUTPUT2_PRIVKEY=<OUTPUT2’s private key>
    export DAO=<dao address>
    ```
3. Make sure all addresses are stored in the keybase and have enough tokens.
4. Submit a non-custodial stake tx to stake a new node.
    ```
    $POKT accounts delete $OUTPUT1
    $POKT nodes stake non-custodial $PUBKEY $OUTPUT1 $STAKE $CHAINS $URL $NETWORK 10000 false
    ```
5. Submit three stake transactions to change the output address signed by the new output address, the operator address, and the current output address respectively.  Verify all transactions are rejected with CodeUnauthorizedSigner (pos:125), CodeUnequalOutputAddr (pos:124), and CodeUnauthorized (sdk:4).
    ```
    $POKT accounts import-raw $OUTPUT1_PRIVKEY
    $POKT nodes stake non-custodial $PUBKEY $OUTPUT2 $STAKE $CHAINS $URL $NETWORK 10000 false
    $POKT accounts delete $OUTPUT2
    $POKT nodes stake non-custodial $PUBKEY $OUTPUT2 $STAKE $CHAINS $URL $NETWORK 10000 false
    $POKT accounts delete $NODE
    $POKT nodes stake non-custodial $PUBKEY $OUTPUT2 $STAKE $CHAINS $URL $NETWORK 10000 false
    ```
6. Enable OEDIT.  Choose the upgrade height accordingly.
    ```
    $POKT gov enable $DAO <upgrade height> OEDIT $NETWORK 10000
    ```
7. Submit three stake transactions to change the output address signed by the new output address, the operator address, and the current output address respectively.  Verify the first two transactions are rejected with CodeUnauthorizedSigner (pos:125) and CodeDisallowedOutputAddressEdit (pos:127) respectively, and the last transaction is accepted.
    ```
    $POKT accounts import-raw $PRIVKEY
    $POKT accounts import-raw $OUTPUT2_PRIVKEY
    $POKT nodes stake non-custodial $PUBKEY $OUTPUT2 $STAKE $CHAINS $URL $NETWORK 10000 false
    $POKT accounts delete $OUTPUT2
    $POKT nodes stake non-custodial $PUBKEY $OUTPUT2 $STAKE $CHAINS $URL $NETWORK 10000 false
    $POKT accounts delete $NODE
    $POKT nodes stake non-custodial $PUBKEY $OUTPUT2 $STAKE $CHAINS $URL $NETWORK 10000 false
    ```
8. Verify the node’s output address is changed to $OUTPUT2.
